### PR TITLE
Add Code for Zero_N_Knapsack

### PR DESCRIPTION
**Pull request by ANUJAGRAWAL2000**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)
Yes I have read the rules(12).
### Description

(Write your answer here.)

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

(Write your answer here.)

### CODE REVIEW

**Comment by ANUJAGRAWAL2000** [comment]
I have added the Code for Zero_N_Knapsack with DP and this code is not already existed.I have also add one input and output example and also the problem statement.Thanks in Adavnce!

**Comment by atarax665** [comment]
Please fill the PR form and add the issue number of the issue that this PR resolves.

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/*
+You've heard of 0-1 knapsack. Below is the problem statement for the same.
```
Comment (atarax665) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]:
```suggestion

```

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+Given weights and values of n items, put these items in a knapsack of capacity cap to get the maximum total value in the knapsack. In other words, given two integer arrays val[0..n-1] and wt[0..n-1] which represent values and weights associated with n items respectively. Also given an integer cap which represents knapsack capacity, find out the maximum value subset of val[] such that sum of the weights of this subset is smaller than or equal to cap. You cannot break an item, either pick the complete item, or don�t pick it (0-1 property).
```
Comment (atarax665) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Please format the description in such a way that the whole description is readable on a phone or a laptop screen at once. And please also add a description of your approach to solve this problem.

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+There is a little twist for 0-N knapsack. You can pick an element more than once. Now you have to find maximum value multi subset of val[] such that sum of the weights of this subset is smaller than or equal to cap.
+
+Note: Maximum value subset means the sunset with maximum sum of all the values in subset.
+
```
Comment (atarax665) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp:42]: Add space and time complexity of your code.

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+3 2 4 5 1
+4 3 5 6 1
+Sample Output
+16
+
+*/
+
+#include<iostream>
+#include<map>
+#include<algorithm>
```
Comment (atarax665) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: You do not need to add any other header file if you use  ```#include<bits/stdc++.h>```.

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+			M->second=V[i];
+		}
+		else
+		{
+			A.insert(make_pair(S[i],V[i]));
+		}
+	}
+    long long int T=A.size();
+    i=0;
+	cout<<KnapSack_2(C,S,V,N);
```
Comment (atarax665) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Inconsistent indentation

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+			}
+		}
+	}
+	return dp[C];
+}
+
+int main()
+{
+	long long int N,C,i,j;
+	cin>>N>>C;
```
Comment (atarax665) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Format your code using some C++ Formatter.

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+*/
+
+#include<iostream>
+#include<map>
+#include<algorithm>
+#include<bits/stdc++.h>
+using namespace std;
+long long int U=0;
+long long int B[2][1005];
+void KnapSack(long long int T,long long int Cap,long long int S,long long int i,long long int V)
```
Comment (atarax665) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Use better variable names instead of alphabets.

**Review by atarax665 — CHANGES_REQUESTED** [review]
Update appropriate README.md

**Comment by ANUJAGRAWAL2000** [comment]
Add space and Time Complexity, Used Better Variables, Format Code using C++ Formatter, Indentation removed, Header Files removed, Mobile acceptable also.

**Comment by HarshCasper** [comment]
Please make the README changes then. I will review the Code then 👍

**Comment by ANUJAGRAWAL2000** [comment]
> Please make the README changes then. I will review the Code then 👍


Update README.md.

**Comment by ANUJAGRAWAL2000** [comment]
Hii!!
I have update all the changes requested. Could u plz look at my PR now.

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
+// Program to print BFS traversal from a given
```
Comment (atarax665) [inline C-Plus-Plus/graphs/BFS1.cpp]: Use milti-line comment here.

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  // Pointer to an array containing adjacency
+  // lists
+  list < int > * adj;
+  public:
+    Graph(int V); // Constructor
+
+  // function to add an edge to graph
+  void addEdge(int v, int w);
+
+  // prints BFS traversal from a given source s
```
Comment (atarax665) [inline C-Plus-Plus/graphs/BFS1.cpp]: Remove all such trivial comments.

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        visited[ * i] = true;
+        queue.push_back( * i);
+      }
+    }
+  }
+}
+
+int main() {
+  // Create a graph given in the above diagram
+  Graph g(4);
```
Comment (atarax665) [inline C-Plus-Plus/graphs/BFS1.cpp]: Dynamic input is required.

**Review by atarax665 — CHANGES_REQUESTED** [review]

File: C-Plus-Plus/graphs/BFS1.cpp (hunk 10 above)
Comment (ANUJAGRAWAL2000) [reply C-Plus-Plus/graphs/BFS1.cpp]: I have used multi line comments, remove trivial comments, and also make the input dynamic accepted by user.
Now plz look at my PR and suggest me if any change required or plz merge it.👍

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+Graph::Graph(int V) {
+  this -> V = V;
+  adj = new list < int > [V];
+}
+
+void Graph::addEdge(int v, int w) {
+  adj[v].push_back(w);
+}
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/BFS1.cpp]: Format the entire Code using [this](http://format.krzaq.cc/)

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  }
+  cout << "Enter the vertex to start BFS\n";
+  cin >> Start_Vertex ;
+  cout << "Following is Breadth First Traversal"
+   <<  "(starting from vertex )" << Start_Vertex 
+   << " \n";
+  g.BFS(Start_Vertex);
+
+  return 0;
+}
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/BFS1.cpp]: Format your Code using [this](http://format.krzaq.cc/)

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+//Time Complexity-O(n)
+//Space Complexity-O(n)
+//where n is the no of nodes in the graph.
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/BFS1.cpp]: Add sample I/O

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
+// Program to print BFS traversal 
+// from a given source vertex. BFS(int s)
+// traverses vertices reachable from s.
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/BFS1.cpp]: Add proper information on what your Code does. Don't give a very vague description

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  cout << KnapSack_2(Capacity_Knapsack, Val, Wt, No_Of_Items);
+  return 0;
+}
+
+/*
+TimeComplexity=O(n*W)
+Auxiliary Space=O(W)
+n=No. of Items
+W=Capacity of Knapsack
+*/
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Add a new line at the EOF

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+4 3 5 6 1
+Sample Output
+16
+*/
+#include<bits/stdc++.h>
+
+using namespace std;
+
+long long int dp[1000] = {
+  0
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Push sample I/O at the bottom of Code

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        M -> second = Wt[i];
+    } else {
+      A.insert(make_pair(Val[i], Wt[i]));
+    }
+  }
+  long long int T = A.size();
+  i = 0;
+  cout << KnapSack_2(Capacity_Knapsack, Val, Wt, No_Of_Items);
+  return 0;
+}
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Format your Code using [this](http://format.krzaq.cc/)

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+smaller than or equal to cap. You cannot 
+break an item, either pick the complete item, 
+or don�t pick it (0-1 property).
+There is a little twist for 0-N knapsack. You
+can pick an element more than once. Now you 
+have to find maximum value multi subset of val[]
+such that sum of the weights of this subset is
+smaller than or equal to cap.
+Note: Maximum value subset means the sunset with
+maximum sum of all the values in subset.
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Make this description compact if possible and more readable.

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by ANUJAGRAWAL2000** [comment]
I have updated both the code as requested. Plz look to the PR and if any change required, inform me and then plz merge it.

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+//referred to as a 'search key'[1]), and
+//explores all of the neighbor nodes at
+//the present depth prior to moving on
+//to the nodes at the next depth level.
+// This Program is to print BFS traversal
+// from a given source vertex. BFS(int s)
+// traverses vertices reachable from s.
+// As Input it required No of Vertex, No of Edges
+// which Edges are connected, Source Vertex.
+//In Output we get the BFS traversal.
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/BFS1.cpp]: Use multi-line comments

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+        g.addEdge(A, B);
+    }
+    cout << "Enter the vertex to start BFS\n";
+    cin >> Start_Vertex;
+    cout << "Following is Breadth First Traversal"
+         << "(starting from vertex )" << Start_Vertex
+         << " \n";
+    g.BFS(Start_Vertex);
+
+    return 0;
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/BFS1.cpp]: Format the Code properly

File: C-Plus-Plus/graphs/BFS1.cpp [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+//6
+//0 1
+//0 2
+//1 2
+//2 0
+//2 3
+//3 3
+//2
+//Output
+//2 0 3 1
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/BFS1.cpp]: Put it in a multi-line comment

File: C-Plus-Plus/graphs/BFS1.cpp (hunk 19 above)
Comment (HarshCasper) [inline C-Plus-Plus/graphs/BFS1.cpp]: Also make the content look presentable by making it minimal and compact

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+(0-1 property).There is a little twist 
+for 0-N knapsack. You can pick an element
+more than once. Now you have to find 
+maximum value multi subset of val[]
+such that sum of the weights of this 
+subset is smaller than or equal to cap.
+Note: Maximum value subset means the 
+subset with maximum sum of all the 
+values in subset.
+*/
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Make the content look presentable by making it minimal and compact

Comment (HarshCasper) [inline C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: Use multi-line comments

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by ANUJAGRAWAL2000** [comment]
Update Both Zero_N_Knapsack.cpp and BFS1.cpp as suggested. Plz look at it and suggest if changes required.
I have make both description minimal and compact.

File: C-Plus-Plus/dp/Zero_N_Knapsack.cpp (hunk 22 above)
Comment (ANUJAGRAWAL2000) [reply C-Plus-Plus/dp/Zero_N_Knapsack.cpp]: It is already multi-line.

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**